File: masar_cortex/custom/delivery_note/delivery_note.js
// Copyright (c) 2023, KCSC and contributors
// For license information, please see license.txt

frappe.ui.form.on('Delivery Note', {
	refresh: function(frm) {
        createGatePass(frm);
	}
});

function createGatePass(frm) {
    if (frm.doc.docstatus === 1) {
        frm.add_custom_button(__("Create Gate Pass"), function() {
                frappe.call({
                    method: "masar_cortex.custom.delivery_note.delivery_note.create_gate_pass",
                    args: {
                        dn_name: frm.doc.name,
                    },
                    callback: function(r) {
                        if (r.message) {
                            frappe.set_route('Form', 'Gate Pass', r.message);
                        }
                    }
                });
            });
    }
}

File: masar_cortex/custom/delivery_note/delivery_note.py
import frappe

@frappe.whitelist()
def create_gate_pass(dn_name):
    if not dn_name:
        frappe.throw("Delivery Note name is required.")

    dn = frappe.get_doc("Delivery Note", dn_name)
    gate_pass = frappe.new_doc("Gate Pass")
    gate_pass.delivery_note = dn.name
    gate_pass.customer = dn.customer
    gate_pass.posting_date = frappe.utils.nowdate()

    gate_pass.insert()
    frappe.db.commit()

    return gate_pass.name